### PR TITLE
[EW] Added basic Jenkins Stub functionality

### PR DESCRIPTION
**Pull request by EWright213**

### CODE REVIEW

File: src/JenkinsStub.kt [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+                post {
+
+                    call.respondText(status = HttpStatusCode.BadRequest) { "" }
+                }
+            }
+
+
+            route("/$jenkinsUrl/job/job-name-no-params/build") {
+                post("/") {
+                    call.respondText(status = HttpStatusCode.OK) { "" }
```
Comment (eledg) [inline src/JenkinsStub.kt:44]: This can just be `call.respond(HttpStatusCode.OK)`

File: src/JenkinsStub.kt [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    val jenkinsUrl = "jenkins-url"
+
+
+    routing {
+        println("Starting routing")
+        authenticate("myBasicAuth") {
+            route("/$jenkinsUrl/job/job-name-with-params/buildWithParameters") {
+                val jobParams = "GOOD_PARAM"
```
Comment (eledg) [inline src/JenkinsStub.kt:28]: Is this variable used?

File: src/JenkinsStub.kt [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        }
+    }
+
+    val jenkinsUrl = "jenkins-url"
+
+
+    routing {
+        println("Starting routing")
+        authenticate("myBasicAuth") {
+            route("/$jenkinsUrl/job/job-name-with-params/buildWithParameters") {
```
Comment (eledg) [inline src/JenkinsStub.kt:27]: You can have multiple routes, so don't need to define `jenkinsUrl` as a variable, e.g.
```
route("jenkins-url") {
    route("job/job-name") {
        // do something
    }
}
```

Having said that, I don't think `jenkins-url` is needed as part of this? Is it part of the api path for a jenkins instance?

**Review by eledg — CHANGES_REQUESTED** [review]

File: src/JenkinsStub.kt (hunk 1 above)
Comment (EWright213) [reply src/JenkinsStub.kt]: Fixed, thanks!

File: src/JenkinsStub.kt (hunk 2 above)
Comment (EWright213) [reply src/JenkinsStub.kt]: Removed, thanks!

File: src/JenkinsStub.kt (hunk 3 above)
Comment (EWright213) [reply src/JenkinsStub.kt]: Tidied up as per your example. Yes, I believe that the jenkins-url is needed as the api path, as per https://www.jenkins.io/doc/book/using/remote-access-api/

Comment (eledg) [reply src/JenkinsStub.kt]: I think in that example `JENKINS_URL` is meant to just be the domain of Jenkins. In the case of this stub locally it'd be `http://localhost:8084`. Could you compare this to a known Jenkins instance?

Comment (EWright213) [reply src/JenkinsStub.kt]: Ah I see, you're right! Fixed, thanks!

File: gradle.properties [hunk 4]
```diff
@@ -0,0 +1,5 @@
+kotlin.code.style=official
+ktor_version=1.4.3
+kotlin_version=1.4.20
+logback_version=1.2.1
```
Comment (eledg) [inline gradle.properties:4]: Are these the latest versions?

File: gradle/wrapper/gradle-wrapper.properties [hunk 5]
```diff
@@ -0,0 +1,5 @@
+distributionBase=GRADLE_USER_HOME
+distributionPath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-6.3-bin.zip
```
Comment (eledg) [inline gradle/wrapper/gradle-wrapper.properties:3]: This version could be updated too

File: src/JenkinsStub.kt [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+fun Application.module(testing: Boolean = false) {
+    install(Authentication) {
+        basic("myBasicAuth") {
+            realm = "Jenkins stub"
+            validate { if (it.name == "username" && it.password == "password") UserIdPrincipal(it.name) else null }
+        }
+    }
+
+    routing {
+        println("Starting routing")
```
Comment (eledg) [inline src/JenkinsStub.kt:22]: Please can the println be removed. If logging is needed here it should be through the logger.

**Review by eledg — CHANGES_REQUESTED** [review]

File: gradle.properties (hunk 4 above)
Comment (EWright213) [reply gradle.properties]: Updated, thanks!

File: gradle/wrapper/gradle-wrapper.properties (hunk 5 above)
Comment (EWright213) [reply gradle/wrapper/gradle-wrapper.properties]: Done, thanks!

File: src/JenkinsStub.kt (hunk 6 above)
Comment (EWright213) [reply src/JenkinsStub.kt]: Removed, I don't think any logging is needed - would you recommend any?

Comment (eledg) [reply src/JenkinsStub.kt]: Nope, I agree

**Review by eledg — APPROVED** [review]

**Pull request merged**